Initial addition of staBal routes.

johngrantuk opened the pull request.

TomAFrench commented on src/routeProposal/filtering.ts:
```diff
@@ -39,8 +41,9 @@ export function filterPoolsOfInterest(
+    chainId: number,
     currentBlockTimestamp = 0
-): [PoolDictionary, string[]] {
+): [PoolDictionary, string[], StablePool] {
```
I'm not a fan of functions which return many different things. Can we pull this StablePool out elsewhere?

TomAFrench commented:
It's not really clear what's going on here, I think we need to add more documentation on what the flow of the algorithm is doing. It looks like we're going to be passing along incomplete paths to later stages of the algorithm.

TomAFrench commented on test/staBalPaths.spec.ts:
```diff
@@ -0,0 +1,355 @@
+// TS_NODE_PROJECT='tsconfig.testing.json' npx mocha -r ts-node/register test/staBalPaths.spec.ts
```
Many of these tests are pretty much copy pastes from each other. Can you deduplicate these where you're just passing in different inputs to the same code?

This is something that we need to be doing generally with this codebase. The amount of duplication in the tests is huge.

johngrantuk replied:
Deduplicated and added more thorough Path testing.

TomAFrench commented on src/routeProposal/filtering.ts:
```diff
@@ -33,15 +35,20 @@ The main purpose of this function is to:
     tokenIn: string,
     tokenOut: string,
     maxPools: number,
+    chainId: number,
```
This is unused.

TomAFrench commented on src/routeProposal/filtering.ts:
```diff
@@ -33,15 +35,20 @@ The main purpose of this function is to:
-    const poolsDictionary: PoolDictionary = {};
+): [PoolDictionary, string[], PoolDictionary] {
+    // This will include all pools
+    const poolsAllDictionary: PoolDictionary = {};
```
Personally I'd create this dictionary outside of this function and then pass it in to be filtered. We're overloading what the purpose of this function is by doing this.


My solution would look something like this:
```
const poolsDict = Object.fromEntries(pools
  .map((pool) => [pool.id, parseNewPool(pool, swapOptions.timestamp)])
  .filter(([,pool])=> pool !== undefined)
)
const [poolsFilteredDict, hopTokens] =
    filterPoolsOfInterest(
        poolsDict,
        tokenIn,
        tokenOut,
        swapOptions.maxPools,
    );
```

johngrantuk replied:
Created a new function `parseToPoolsDict`. This also filters out pools with 0 balance.

TomAFrench commented on src/constants.ts:
```diff
@@ -25,6 +25,48 @@ export const VAULTADDR: { [chainId: number]: string } = {
+export const USDCCONNECTINGPOOL: {
+    [chainId: number]: { id: string; usdc: string };
+} = {
+    1: {
```
Lets remove these keys if we're not using them

TomAFrench reviewed (changes requested):
I've pushed a couple of changes to mostly rename some variables/comments and have better return values for some functions.

I think we're still overloading `filterPoolsOfInterest` however and the tests can be deduplicated. Happy with the general algo though.

TomAFrench commented on test/staBalPaths.spec.ts:
```diff
@@ -0,0 +1,355 @@
+        const tokenIn = '0x0000000000085d4780B73119b644AE5ecd22b376';
+        const tokenOut = '0x9a71012b13ca4d3d0cdc72a177df3ef03b0e76a3'; // BAL
+
+        const testPools: any = cloneDeep(staBalPools.pools);
```
Why are these being converted to `any`?

TomAFrench reviewed (approved):
~~LGTM. @johngrantuk can you check you're happy with these final changes?~~

Ah, forgot you're improving some of the tests, nvm.

johngrantuk commented:
@TomAFrench I've changed the previously highlighted tests to use all pools. Also updated where required to make things work after the removal of BigNumber.js. Would probably benefit from another quick check.

`constants.ts` needs updated with `USDCCONNECTINGPOOL` and `STABALADDR` information once those pools are deployed on Polygon so not sure if we merge before or after that?

TomAFrench commented:
> I've changed the previously highlighted tests to use all pools. Also updated where required to make things work after the removal of BigNumber.js. Would probably benefit from another quick check.

Thanks, I'll give this a check over in a bit and then merge.

> constants.ts needs updated with USDCCONNECTINGPOOL and STABALADDR information once those pools are deployed on Polygon so not sure if we merge before or after that?

This isn't an issue. We just need to remove the entries for each networks and we'll fall straight through the new function so we can merge now.
